cpp/src/arrow/filesystem/hdfs_test.cc

```diff
@@ -0,0 +1,169 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <memory>
+#include <string>
+
+#include <gtest/gtest.h>
+
+#include "arrow/filesystem/hdfs.h"
+#include "arrow/filesystem/test_util.h"
+#include "arrow/testing/gtest_util.h"
+#include "arrow/util/io_util.h"
+
+namespace arrow {
+namespace fs {
+namespace internal {
+
+#ifndef _WIN32
+
+#define SOME_DATA_SIZE 9
+
+using HdfsDriver = arrow::io::HdfsDriver;
+
+struct JNIDriver {
+  static HdfsDriver type;
+};
+
+struct PivotalDriver {
+  static HdfsDriver type;
+};
+
+template <typename DRIVER>
+class TestHadoopFileSystem : public ::testing::Test {
+ public:
+  void SetUp() override {
+    const char* host = std::getenv("ARROW_HDFS_TEST_HOST");
+    const char* port = std::getenv("ARROW_HDFS_TEST_PORT");
+    const char* user = std::getenv("ARROW_HDFS_TEST_USER");
+
+    std::string hdfs_host = host == nullptr ? "localhost" : std::string(host);
+    int hdfs_port = port == nullptr ? 20500 : atoi(port);
+    std::string hdfs_user = user == nullptr ? "root" : std::string(user);
+
+    bool use_hdfs3;
+    if (DRIVER::type == HdfsDriver::LIBHDFS) {
+      use_hdfs3 = false;
+    } else {
+      use_hdfs3 = true;
+    }
+
+    HdfsOptions hdfs_options;
+    hdfs_options.ConfigureEndPoint(hdfs_host, hdfs_port);
+    hdfs_options.ConfigureHdfsUser(hdfs_user);
+    hdfs_options.ConfigureHdfsDriver(use_hdfs3);
+    hdfs_options.ConfigureHdfsReplication(0);
+
+    std::shared_ptr<HadoopFileSystem> hdfs;
+    auto status = HadoopFileSystem::Make(hdfs_options, &hdfs);
+    if (!status.ok()) {
+      loaded_driver_ = false;
+      return;
+    }
+    loaded_driver_ = true;
+    fs_ = std::make_shared<SubTreeFileSystem>("", hdfs);
```

Comment:
What is the use of `SubTreeFileSystem` here?

Reply:
I followed this way in localfs_test.cc and s3fs_test.cc, I suppose this is desired, to make a filesystem with speific type and then use SubTreeFileSystem to make it general.

Reply:
`SubTreeFileSystem` is only used if it's needed to run the test from a subdirectory (for example for local filesystem tests: you don't want tests to write at the filesystem root). This isn't the case here.

Reply:
in the latest commit, make subTreeFikeSystem is removed, not sure why this line here is not updated.